src/pydvl/influence/torch/influence_function_model.py

```diff
@@ -876,3 +881,322 @@ def to(self, device: torch.device):
+        for key in self.active_layers.keys():
+            evals_a, evecs_a = torch.linalg.eigh(forward_x[key])
+            evals_g, evecs_g = torch.linalg.eigh(grad_y[key])
+            layers_evecs_a[key] = evecs_a
+            layers_evect_g[key] = evecs_g
+            layers_diags[key] = torch.kron(evals_g.view(-1, 1), evals_a.view(-1, 1))
+
+        self.ekfac_representation = EkfacRepresentation(
+            self.active_layers.keys(),
+            self.active_layers.values(),
+            layers_evecs_a.values(),
+            layers_evect_g.values(),
+            layers_diags.values(),
+        )
+        return self
+
+    @staticmethod
+    def _init_layer_diag(module: torch.nn.Module) -> torch.Tensor:
+        """
+        Initialize the tensor that will store the updated diagonal values of the layer.
+        """
+        if isinstance(module, nn.Linear):
+            with_bias = module.bias is not None
+            sG = module.out_features
+            sA = module.in_features + int(with_bias)
+            layer_diag = torch.zeros((sA * sG), device=module.weight.device)
+        else:
+            raise NotImplementedLayerRepresentationException(
+                f"Only Linear layers are supported, but found module {module} requiring grad."
+            )
+        return layer_diag
+
+    def _get_layer_diag_hooks(
+        self,
+        m_name: str,
+        module: torch.nn.Module,
+        last_x_kfe: Dict[str, torch.Tensor],
+        diags: Dict[str, torch.Tensor],
+    ) -> Tuple[Callable, Callable]:
+        """
+        Create the hooks that will be used to update the diagonal values of the layer.
+        The hooks are registered to the layer and will be called during the forward and
+        backward passes. At each pass, the hooks will update the tensor that stores the
+        updated diagonal values of the layer. This tensor is stored in the diags
+        dictionary.
+        """
+        evecs_a, evecs_g = self.ekfac_representation.get_layer_evecs()
+        if isinstance(module, nn.Linear):
+            with_bias = module.bias is not None
+
+            def input_hook(m, x, y):
+                x = x[0]
+                if with_bias:
+                    x = torch.cat(
+                        (x, torch.ones((x.shape[0], 1), device=module.weight.device)),
+                        dim=1,
+                    )
+                last_x_kfe[m_name] = torch.mm(x, evecs_a[m_name])
+
+            def grad_hook(m, m_grad, m_out):
+                m_out = m_out[0]
+                gy_kfe = torch.mm(m_out, evecs_g[m_name])
+                diags[m_name] += torch.mm(
+                    gy_kfe.t() ** 2, last_x_kfe[m_name] ** 2
+                ).view(-1)
+
+        else:
+            raise NotImplementedLayerRepresentationException(
+                f"Only Linear layers are supported, but found module {module} requiring grad."
+            )
+        return input_hook, grad_hook
+
+    def update_diag(
+        self,
```

Comment:
Still, having this separate method could be avoided. Is my understanding correct, that this is the eigenvalue correction (so the E in Ekfac?). I would suggest to introduce an init parameter, if the kfac blocks should be adapted to this update logic. What do you think?